src/hotspot/share/gc/shenandoah/shenandoahOldGC.cpp

```diff
@@ -84,3 +124,50 @@ bool ShenandoahOldGC::collect(GCCause::Cause cause) {
   entry_rendezvous_roots();
   return true;
 }
+
+void ShenandoahOldGC::entry_coalesce_and_fill_message(char *buf, size_t len) const {
+  // ShenandoahHeap* const heap = ShenandoahHeap::heap();
+  jio_snprintf(buf, len, "Coalescing and filling (%s)", _generation->name());
+}
+
+void ShenandoahOldGC::op_coalesce_and_fill() {
+  ShenandoahHeap* const heap = ShenandoahHeap::heap();
+
+  WorkGang* workers = heap->workers();
+  uint nworkers = workers->active_workers();
+
+  assert(_generation->generation_mode() == OLD, "Only old-GC does coalesce and fill");
+
+  ShenandoahHeuristics* old_heuristics = heap->old_generation()->heuristics();
+  uint coalesce_and_fill_regions_count = old_heuristics->old_coalesce_and_fill_candidates();
+  ShenandoahHeapRegion* coalesce_and_fill_region_array[coalesce_and_fill_regions_count];
+
+  old_heuristics->get_coalesce_and_fill_candidates(coalesce_and_fill_region_array);
+  ShenandoahConcurrentCoalesceAndFillTask task(nworkers, coalesce_and_fill_region_array, coalesce_and_fill_regions_count);
+
+
+  // TODO:  We need to implement preemption of coalesce and fill.  If young-gen wants to run while we're working on this,
+  // we should preempt this code and then resume it after young-gen has finished.  This requires that we "remember" the state
+  // of each worker thread so it can be resumed where it left off.  Note that some worker threads may have processed more regions
+  // than others at the time of preemption.
+
+  workers->run_task(&task);
+}
+
+void ShenandoahOldGC::entry_coalesce_and_fill() {
+  char msg[1024];
+  ShenandoahHeap* const heap = ShenandoahHeap::heap();
+
+  entry_coalesce_and_fill_message(msg, sizeof(msg));
+  ShenandoahConcurrentPhase gc_phase(msg, ShenandoahPhaseTimings::coalesce_and_fill);
+
+  // Not sure if I want/need any of the following here:
+  //
+  // TraceCollectorStats tcs(heap->monitoring_support()->concurrent_collection_counters());
+  // EventMark em("%s", msg);
```

Comment:
Yes, I think we want them.

Reply:
I've uncommented this so it is now included.